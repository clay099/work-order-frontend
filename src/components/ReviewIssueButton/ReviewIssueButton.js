import React from "react";
import { useSelector } from "react-redux";
import { Button, TextField } from "@material-ui/core/";
import AdvancedDialog from "../AdvancedDialog/AdvancedDialog";
import useToggle from "../../hooks/useToggle";
import useFields from "../../hooks/useFields";
import { submitIssueWithAPI } from "../../actions/projects";
import { useDispatch } from "react-redux";
import useStyles from "./styles";

/** ReviewIssueButton Component
 * @param  {int} id
 *
 * Gets token from Redux state
 *
 * Creates formData, handleChange, resetFormData from useFields custom hook to allow for form inputs and changes
 *
 * Creates a handleSubmit function to submit a completed issues form
 *
 * Renders:
 *    - Button - when clicked opens a Dialog with issues form
 *    - AdvancedDialog Component - displays issues form along with the function to submit the form
 */
const ReviewIssueButton = ({ id }) => {
	const classes = useStyles();
	const dispatch = useDispatch();

	const { token } = useSelector((st) => ({
		token: st.login.token,
	}));

	const {
		setTrue: handleOpen,
		setFalse: handleClose,
		toggle: open,
	} = useToggle(false);

	const { formData, handleChange, resetFormData } = useFields({
		issues: "",
	});

	let FormData = [
		<TextField
			required
			id="issues"
			label="Issues"
			name="issues"
			value={formData.issues}
			onChange={handleChange}
			autoFocus={true}
			fullWidth={true}
		/>,
	];

	const handleSubmit = async (e) => {
		e.preventDefault();

		await dispatch(
			submitIssueWithAPI({
				issues: formData.issues,
				token,
				projectId: id,
			})
		);

		resetFormData();
		handleClose();
	};

	return (
		<>
			<Button
				variant="outlined"
				color="primary"
				onClick={handleOpen}
				className={classes.styledButton}
			>
				Add Project Issue
			</Button>
			<AdvancedDialog
				handleClose={handleClose}
				open={open}
				handleSubmit={handleSubmit}
				titleText="Add Project Issue"
				id="projectIssue"
				buttonText="Submit Issue"
				buttonColor="primary"
				FormData={FormData}
				handleChange={handleChange}
				maxWidth="sm"
			/>
		</>
	);
};

export default ReviewIssueButton;
