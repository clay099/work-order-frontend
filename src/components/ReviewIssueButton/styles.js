import { makeStyles } from "@material-ui/core/styles";
import { styledButton } from "../../styles/styleObjects";

const useStyles = makeStyles((theme) => {
	return styledButton(theme);
});

export default useStyles;
