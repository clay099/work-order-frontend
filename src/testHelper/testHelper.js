// testHelper.js
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
// Import your own reducer
import reducer from "../reducers/rootReducer";

function render(
	ui,
	{
		initialState,
		store = createStore(reducer, initialState),
		...renderOptions
	} = {}
) {
	function Wrapper({ children }) {
		return <Provider store={store}>{children}</Provider>;
	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const tokenState = { token: "token" };
const userTypeUserState = { userType: "user" };
const userTypeTradesmenState = { userType: "user" };
const idState = { id: 1234 };
const userDetailsState = { userDetails: "userDetails" };
const handleSubmitState = { handleSubmit: jest.fn() };
const handleChangeState = { handleChange: jest.fn() };
const disabledState = { disabled: false };
const disabledEmailState = { disabledEmail: false };
const buttonTextState = { buttonText: "buttonText" };
const formDataState = {
	formData: {
		id: 1,
		firstName: "firstName",
		lastName: "lastName",
		email: "email",
		password: "password",
		newPassword: "newPassword",
		confirmNewPassword: "confirmNewPassword",
	},
};

// re-export everything
export * from "@testing-library/react";
// override render method
export {
	render,
	tokenState,
	userTypeUserState,
	userTypeTradesmenState,
	idState,
	userDetailsState,
	handleSubmitState,
	handleChangeState,
	disabledState,
	disabledEmailState,
	buttonTextState,
	formDataState,
};
