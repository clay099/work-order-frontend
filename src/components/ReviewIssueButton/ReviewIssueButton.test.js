import React from "react";
import ReviewIssueButton from "./ReviewIssueButton";
import { render } from "../../testHelper/testHelper";
console.error = jest.fn();

const params = {
	id: 1,
};

const initialState = {
	login: { token: "token" },
};

describe("ReviewIssueButton Component", () => {
	// smoke test
	test("renders without crashing", () => {
		render(<ReviewIssueButton {...params} />, {
			initialState,
		});
	});

	// snapshot test
	test("matches snapshot", () => {
		const { asFragment } = render(<ReviewIssueButton {...params} />, {
			initialState,
		});
		expect(asFragment).toMatchSnapshot();
	});

	test("value to be in the document", () => {
		const { getByText } = render(<ReviewIssueButton {...params} />, {
			initialState,
		});
		const value = getByText("Add Project Issue");
		expect(value).toBeInTheDocument();
	});
});
