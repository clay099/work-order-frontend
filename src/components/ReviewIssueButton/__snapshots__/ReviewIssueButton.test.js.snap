// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewIssueButton Component matches snapshot 1`] = `[Function]`;
